batch2/batch/driver.py

```diff
@@ -0,0 +1,470 @@
+
+    async def read_pod_log(self, container):
+        assert container in tasks
+        log.info(f'reading container log for {self.name}, {container} from instance {self.instance}')
+
+        if self.instance is None:
+            return None, None
+
+        inst = self.instance
+        url = f'http://{inst.ip_address}:5000/api/v1alpha/pods/{self.name}/containers/{container}/log'
+        resp, err = self._get(url)
+        if resp:
+            return resp.json(), None
+        else:
+            assert err
+            log.info(f'failed to read container log {self.name}, {container} on {inst} due to err {err}, ignoring')
+            return None, err
+
+    async def read_container_status(self, container):
+        assert container in tasks
+        log.info(f'reading container status for {self.name}, {container} from instance {self.instance}')
+
+        if self.instance is None:
+            return None, None
+
+        inst = self.instance
+        url = f'http://{inst.ip_address}:5000/api/v1alpha/pods/{self.name}/containers/{container}/status'
+        resp, err = self._get(url)
+        if resp:
+            return resp.json(), None
+        else:
+            assert err
+            log.info(f'failed to read container status {self.name}, {container} on {inst} due to err {err}, ignoring')
+            return None, err
+
+    def status(self):
+        if self._status is None:
+            return {
+                'metadata': self.spec['metadata'],
+                'status': {
+                    'containerStatuses': None,
+                    'phase': 'Pending'
+                }
+            }
+        else:
+            return self._status
+
+    def __str__(self):
+        return self.name
+
+
+class Driver:
+    def __init__(self, k8s, batch_bucket, batch_gsa_key=None):
+        self.k8s = k8s
+        self.batch_bucket = batch_bucket
+        self.pods = None  # populated in run
+        self.complete_queue = asyncio.Queue()
+        self.ready_queue = asyncio.Queue(maxsize=1000)
+        self.ready = sortedcontainers.SortedSet(key=lambda pod: pod.cores)
+        self.ready_cores = 0
+        self.changed = asyncio.Event()
+
+        self.pool = None  # created in run
+
+        deploy_config = get_deploy_config()
+
+        self.base_url = deploy_config.base_url('batch2')
+
+        self.inst_pool = InstancePool(self)
+
+        if batch_gsa_key is None:
+            batch_gsa_key = os.environ.get('BATCH_GSA_KEY', '/batch-gsa-key/privateKeyData')
+        credentials = google.oauth2.service_account.Credentials.from_service_account_file(batch_gsa_key)
+
+        self.gservices = GServices(self.inst_pool.machine_name_prefix, credentials)
+
+    async def activate_worker(self, request):
+        body = await request.json()
+        inst_token = body['inst_token']
+        ip_address = body['ip_address']
+
+        inst = self.inst_pool.token_inst.get(inst_token)
```

Comment:
I think we should prefer explanatory names so `token_inst.get(x)` is probably better served by a function on `InstancePool` called `get_by_token(x)`. If we name `inst_pool` `instances`, then it looks quite nice: `self.instances.get_by_token(token)`

Reply:
Despite the cuteness of this, I prefer `instance_pool` to `instances`, as the former makes the dynamic nature of the pool clearer.  `self.intance_pool.get_instance_by_token(token)` seems the height of clarity.